src/approaches/gnn_policy_approach.py

```diff
@@ -0,0 +1,497 @@
+        save_path = utils.get_approach_save_path_str()
+        with open(f"{save_path}_{online_learning_cycle}.gnn", "rb") as f:
+            info = pkl.load(f)
+        # Initialize fields from loaded dictionary.
+        input_example, target_example = info["exemplar"]
+        dataset = GraphDictDataset(  # type: ignore
+            [input_example], [target_example])
+        self._gnn = setup_graph_net(  # type: ignore
+            dataset,
+            use_gpu=False,
+            num_steps=CFG.gnn_policy_num_message_passing)
+        self._gnn.load_state_dict(info["state_dict"])
+        self._nullary_predicates = info["nullary_predicates"]
+        self._max_option_objects = info["max_option_objects"]
+        self._max_option_params = info["max_option_params"]
+        self._node_feature_to_index = info["node_feature_to_index"]
+        self._edge_feature_to_index = info["edge_feature_to_index"]
+        self._input_normalizers = info["input_normalizers"]
+        self._target_normalizers = info["target_normalizers"]
+
+    def _predict_option(self, state: State, atoms: Set[GroundAtom],
+                        goal: Set[GroundAtom]) -> _Option:
+        # Get output graph.
+        in_graph, object_to_node = self._graphify_single_input(
+            state, atoms, goal)
+        node_to_object = {v: k for k, v in object_to_node.items()}
+        type_to_obj = defaultdict(set)
+        for obj, node in object_to_node.items():
+            type_to_obj[obj.type.name].add(node)
+        if CFG.gnn_policy_do_normalization:
+            in_graph = normalize_graph(  # type: ignore
+                in_graph, self._input_normalizers)
+        out_graph = get_single_model_prediction(  # type: ignore
+            self._gnn, in_graph)
+        if CFG.gnn_policy_do_normalization:
+            out_graph = normalize_graph(  # type: ignore
+                out_graph,
+                self._target_normalizers,
+                invert=True)
+        # Extract parameterized option and continuous parameters.
+        onehot_output, params = np.split(  # type: ignore
+            out_graph["globals"], [len(self._sorted_options)])
+        param_opt = self._sorted_options[np.argmax(onehot_output)]
+        # Pad and clip parameters.
+        params = params[:param_opt.params_space.shape[0]]
+        params = params.clip(param_opt.params_space.low,
+                             param_opt.params_space.high)
+        # Extract objects, making sure to enforce the typing constraints.
+        objects = []
+        for i, obj_type in enumerate(param_opt.types):
+            scores = out_graph["nodes"][:, i]
+            allowed_idxs = type_to_obj[obj_type.name]
+            for j in range(len(scores)):
+                if j not in allowed_idxs:
+                    scores[j] = float("-inf")  # set its score to be really bad
+            if np.max(scores) == float("-inf"):  # type: ignore
+                # If all scores are None, we failed to select an object.
```

Comment:
is this actually possible? also instead of None, you mean -inf, right?

Reply:
yeah it was happening in a couple episodes in tools, i didn't investigate much bc it was rare enough